cmd/thermal-recorder/throttledRecorder.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"log"
+
+	"github.com/TheCacophonyProject/lepton3"
+)
+
+type ThrottledRecorder struct {
+	recorder              Recorder
+	mainBucket            TokenBucket
+	sparseBucket          TokenBucket
+	recording             bool
+	minRecordingLength    uint32
+	sparseRecordingLength uint32
+	askedToWriteFrame     bool
+	throttledFrames       uint32
+	frameCount            uint32
+}
+
+func NewThrottledRecorder(baseRecorder Recorder, config *ThrottlerConfig, minSeconds int) *ThrottledRecorder {
+	sparseFrames := config.SparseLength * framesHz
+	minFrames := uint16(minSeconds) * framesHz
+
+	if sparseFrames > 0 && sparseFrames < minFrames {
+		sparseFrames = minFrames
+	}
+
+	mainBucketSize := uint32(config.ThrottleAfter * framesHz)
+	supBucketSize := uint32(config.SparseAfter * framesHz)
+	return &ThrottledRecorder{
+		recorder:              baseRecorder,
+		mainBucket:            TokenBucket{tokens: mainBucketSize, size: mainBucketSize},
+		sparseBucket:          TokenBucket{size: supBucketSize},
+		minRecordingLength:    uint32(minFrames),
+		sparseRecordingLength: uint32(sparseFrames),
+	}
+}
+
+type TokenBucket struct {
+	tokens uint32
+	size   uint32
+}
+
+func (bucket *TokenBucket) AddTokens(newTokens uint32) {
+	bucket.tokens += newTokens
+	if bucket.tokens > bucket.size {
+		bucket.tokens = bucket.size
+	}
+}
+
+func (bucket *TokenBucket) RemoveTokens(oldTokens uint32) {
+	if bucket.tokens >= oldTokens {
+		bucket.tokens -= oldTokens
+	} else {
+		bucket.tokens = 0
+	}
+}
+
+func (bucket *TokenBucket) HasTokens(tokens uint32) bool {
+	return bucket.tokens >= tokens
+}
+
+func (bucket *TokenBucket) Empty() {
+	bucket.tokens = 0
+}
+
+func (bucket *TokenBucket) IsFull() bool {
+	return bucket.HasTokens(bucket.size)
+}
+
+func (throttler *ThrottledRecorder) NewFrame() {
```

Comment:
I found the NewFrame confusing as "New" usually means that something is being created. How about `Advance`, `AdvanceFrame`, `Tick` or `NextFrame`?